zkevm-circuits/src/circuit_tools/memory.rs

```diff
@@ -6,77 +6,89 @@ use halo2_proofs::{
     plonk::{Advice, Column, ConstraintSystem, Error, Expression},
     poly::Rotation,
 };
-use std::ops::{Index, IndexMut};
+use std::{
+    collections::HashMap,
+    marker::PhantomData,
+    ops::{Index, IndexMut},
+};
 
 use super::{
-    cached_region::CachedRegion, cell_manager::CellType, constraint_builder::ConstraintBuilder,
+    cached_region::CachedRegion,
+    cell_manager::{CellManager, CellType},
+    constraint_builder::ConstraintBuilder,
 };
 
 #[derive(Clone, Debug, Default)]
-pub(crate) struct Memory<F, C> {
-    columns: Vec<Column<Advice>>,
-    banks: Vec<MemoryBank<F, C>>,
+pub(crate) struct Memory<F: Field, C: CellType, MB: MemoryBank<F, C>> {
+    banks: HashMap<C, MB>,
+    _phantom: PhantomData<F>,
+    tag_counter: usize,
+}
+
+impl<F: Field, C: CellType, MB: MemoryBank<F, C>> Index<C> for Memory<F, C, MB> {
+    type Output = MB;
+
+    fn index(&self, tag: C) -> &Self::Output {
+        if let Some(bank) = self.banks.get(&tag) {
+            bank
+        } else {
+            unreachable!()
+        }
+    }
 }
 
-impl<F: Field, C: CellType> Memory<F, C> {
-    pub(crate) fn new(columns: Vec<Column<Advice>>) -> Self {
+impl<F: Field, C: CellType, MB: MemoryBank<F, C>> IndexMut<C> for Memory<F, C, MB> {
+    fn index_mut(&mut self, tag: C) -> &mut Self::Output {
+        if let Some(bank) = self.banks.get_mut(&tag) {
+            bank
+        } else {
+            unreachable!()
+        }
+    }
+}
+
+impl<F: Field, C: CellType, MB: MemoryBank<F, C>> Memory<F, C, MB> {
+    pub(crate) fn new() -> Self {
         Self {
-            columns,
-            banks: Vec::new(),
+            banks: HashMap::new(),
+            _phantom: PhantomData,
+            tag_counter: 0,
         }
     }
 
-    pub(crate) fn allocate(
+    pub(crate) fn add_rw(
         &mut self,
         meta: &mut ConstraintSystem<F>,
+        cb: &mut ConstraintBuilder<F, C>,
+        cm: &mut CellManager<F, C>,
         tag: C,
-        table_tag: C,
-    ) -> &MemoryBank<F, C> {
-        self.banks.push(MemoryBank::new(
-            meta,
-            self.columns[self.banks.len()],
-            tag,
-            table_tag,
-        ));
-        self.banks.last().unwrap()
+        phase: u8,
+    ) -> &MB {
+        let table_tag = self.allocate_tag();
```

Comment:
If we only ask for one `tag` for each RW pair, why don't we just use the same tag for both side? The lookup is initiated immediately in `MB::new(..)` anyways, so there's no need for extra tags right? Or maybe I'm missing its other usage?

Reply:
We still have to have two to differentiate on the cell manager side in which column to store the inputs (loading) and where to store the table values (stores) right?

Reply:
In this case i feel like it doesn't use the tag at all tho, cuz you can 1) [init two cols with the same tag](https://github.com/CeciliaZ030/zkevm-circuits/blob/813426ba75413e49463b1b6fdad925655fe3395a/zkevm-circuits/src/circuit_tools/memory.rs#L182) 2) [adds it as (data expressions) -> (table expressions)](https://github.com/CeciliaZ030/zkevm-circuits/blob/813426ba75413e49463b1b6fdad925655fe3395a/zkevm-circuits/src/circuit_tools/constraint_builder.rs#L447) 3) [build lookup which performs meta.lookup_any directly on the stored expressions](https://github.com/CeciliaZ030/zkevm-circuits/blob/813426ba75413e49463b1b6fdad925655fe3395a/zkevm-circuits/src/circuit_tools/constraint_builder.rs#L371). For the CM it just needs to know these two cols are for this specific MB that's fine right?

Reply:
You still have to load and store to the different columns, which goes through the cell manager:
- store: https://github.com/CeciliaZ030/zkevm-circuits/blob/fa3ac00f3e83deecf31d991bc5d75fde4d941637/zkevm-circuits/src/circuit_tools/memory.rs#L223-L227
- load: https://github.com/CeciliaZ030/zkevm-circuits/blob/fa3ac00f3e83deecf31d991bc5d75fde4d941637/zkevm-circuits/src/circuit_tools/memory.rs#L239-L243

Without different tags, the cell manager could store these values in any of the two columns, but they really need to be stored in separate columns as the lookup expects.